### PR TITLE
French g2p with pronunciation dictionary

### PR DESCRIPTION
**Pull request by mgrafu**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py [hunk 1]
```diff
@@ -77,6 +84,14 @@
         'ɒ', 'ɔ', 'ə', 'ɛ', 'ɜ', 'ɡ', 'ɪ', 'ɹ', 'ɾ', 'ʃ',
         'ʊ', 'ʌ', 'ʒ', '̃', 'θ'
     ),
+    "fr-FR": (
+        'a', 'b', 'd', 'e', 'f', 'h', 'i', 'j', 'k', 'l', 
+        'm', 'n', 'o', 'p', 'r', 's', 't', 'u', 'v', 'w', 
+        'y', 'z', 'ð', 'ø', 'ŋ', 'œ', 'ɐ', 'ɑ', 'ɒ', 'ɔ', 
+        'ə', 'ɛ', 'ɜ', 'ɡ', 'ɪ', 'ɲ', 'ɹ', 'ʁ', 'ʃ', 'ʊ', 
+        'ʌ', 'ʒ', 'θ',
+        # 'ː', '̃',
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:93]: do we want to keep them or not?

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py [hunk 2]
```diff
@@ -187,6 +202,8 @@ def get_ipa_punctuation_list(locale):
     elif locale == "es-ES":
         # ref: https://en.wikipedia.org/wiki/Spanish_orthography#Punctuation
         punct_set.update(['¿', '¡'])
+    elif locale == "fr-FR":
+        punct_set.update(['–', '“', '”', '…', '̀', '́', '̂', '̈', '̧'])
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:206]: nit: do you have any reference on the punctuation set, like wikipedia or somewhere else?

**Comment by XuesongYang** [comment]
LGTM. Could you please also add some unit tests for the tokenizers?

* https://github.com/NVIDIA/NeMo/blob/main/tests/collections/common/tokenizers/text_to_speech/test_tokenizer_utils.py
* https://github.com/NVIDIA/NeMo/blob/main/tests/collections/common/tokenizers/text_to_speech/test_tts_tokenizers.py

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py (hunk 2 above)
Comment (mgrafu) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:206]: i extracted them from a large French TTS dataset we have, so no external reference for them!

Comment (XuesongYang) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:206]: gotcha! thanks.

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py (hunk 1 above)
Comment (mgrafu) [reply nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py]: resolved offline

File: tests/collections/common/tokenizers/text_to_speech/test_tokenizer_utils.py [hunk 3]
```diff
@@ -120,3 +121,30 @@ def test_any_locale_word_tokenize_with_numbers(self):
...
+        assert output == expected_output
+
+    @pytest.mark.run_only_on('CPU')
+    @pytest.mark.unit
+    def test_french_text_preprocessing_upper(self):
+        input_text = "A BIENTÔT."
+        expected_output = "a bientôt."
+
+        output = french_text_preprocessing(input_text)
+        assert output == expected_output
```
Comment (XuesongYang) [inline tests/collections/common/tokenizers/text_to_speech/test_tokenizer_utils.py:150]: these tests are for `text_preprocessing` funcs, not for any tokenize funcs. Could you pls revise following the above unit tests?

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py [hunk 4]
```diff
@@ -268,6 +269,35 @@ def __init__(
...
+        Args:
+            punct: Whether to reserve grapheme for basic punctuation or not.
+            apostrophe: Whether to use apostrophe or not.
+            add_blank_at: Add blank to labels in the specified order ("last") or after tokens (any non None),
+            if None then no blank in labels.
+            pad_with_space: Whether to pad text with spaces at the beginning and at the end or not.
+            non_default_punct_list: List of punctuation marks which will be used instead default.
+        """
+
+        fr_alphabet = get_grapheme_character_set(locale="fr-FR", case="mixed")
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:289]: when you specified `case="mixed"`, meaning the tokenizer will distinguish uppercase and lowercase letters, and assign different token index. For example, we should expect the final char alphabet like `A-Za-z`. But from your test cases for French, I feel all are lowercase. Could you please double check either you want all lowercased, uppercased, or mixed?

Comment (mgrafu) [reply nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py]: corrected this in newer commit :)

File: nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py [hunk 5]
```diff
@@ -190,6 +204,8 @@ def get_ipa_punctuation_list(locale):
...
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/ipa_lexicon.py:208]: Regarding the extra punctuations used in French, could you pls add comments of unicode for each punctuation just as what "de-DE" did? It is not easy to pinpoint the difference between similar surfaces, such as below. Thanks!       
```
'‒',  # figure dash, U+2012, decimal 8210
'–',  # en dash, U+2013, decimal 8211
```

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py [hunk 6]
```diff
@@ -619,7 +649,7 @@ def __init__(
         Args:
             g2p: Grapheme to phoneme module, should be IpaG2p or some subclass thereof.
             locale: Locale used to determine default text processing logic and punctuation.
-                Supports ["en-US", "de-DE", "es-ES"]. Defaults to "en-US".
+                Supports ["en-US", "de-DE", "es-ES", "fr-FR"]. Defaults to "en-US".
```
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:652]: When adding new language support for `IPATokenizer`, I would suggest to go over its `__init__` function, especially Line 689, where `any_locale_text_preprocessing` is applied for all non en-US locales. `any_locale_text_preprocessing` will preserve the original text's cases. But it seems French needs all lowercased?

File: tests/collections/common/tokenizers/text_to_speech/test_tts_tokenizers.py [hunk 7]
```diff
@@ -118,6 +119,18 @@ def test_spanish_chars_tokenizer(self):
...
+    def test_french_chars_tokenizer(self):
+        input_text = "Bon après-midi !"
+        expected_output = "bon après-midi !"
+
+        tokenizer = FrenchCharsTokenizer()
+        chars, tokens = self._parse_text(tokenizer, input_text)
+
+        assert chars == expected_output
+        assert len(tokens) == len(input_text)
+
```
Comment (XuesongYang) [inline tests/collections/common/tokenizers/text_to_speech/test_tts_tokenizers.py:133]: Could you pls also add a unit test `test_ipa_tokenizer_fr_fr` following the one for de-de below?

```python
    @pytest.mark.run_only_on('CPU')
    @pytest.mark.unit
    def test_ipa_tokenizer_de_de(self):
        input_text = "Hallo welt"
        expected_output = "hˈaloː vˈɛlt"

        g2p = IpaG2p(phoneme_dict=self.PHONEME_DICT_DE, locale="de-DE")
        tokenizer = IPATokenizer(g2p=g2p, locale="de-DE")
        chars, tokens = self._parse_text(tokenizer, input_text)

        assert chars == expected_output
```

File: tests/collections/common/tokenizers/text_to_speech/test_tokenizer_utils.py (hunk 3 above)
Comment (XuesongYang) [reply tests/collections/common/tokenizers/text_to_speech/test_tokenizer_utils.py:150]: we can remove the above three unit tests, and instead add test examples inside all functions related to `any_locale_word_tokenize`. For example, you could extend below `input_text` and `expected_output` as a list by adding fr-fr examples.
```python
@pytest.mark.run_only_on('CPU')
    @pytest.mark.unit
    def test_any_locale_word_tokenize(self):
        input_text = "apple banana pear"
        expected_output = self._create_expected_output(["apple", " ", "banana", " ", "pear"])

        output = any_locale_word_tokenize(input_text)
        assert output == expected_output

    @pytest.mark.run_only_on('CPU')
    @pytest.mark.unit
    def test_any_locale_word_tokenize_with_accents(self):
        input_text = "The naïve piñata at the café..."
        expected_output = self._create_expected_output(
            ["The", " ", "naïve", " ", "piñata", " ", "at", " ", "the", " ", "café", "..."]
        )

        output = any_locale_word_tokenize(input_text)
        assert output == expected_output

    @pytest.mark.run_only_on('CPU')
    @pytest.mark.unit
    def test_any_locale_word_tokenize_with_numbers(self):
        input_text = r"Three times× four^teen ÷divided by [movies] on \slash."
        expected_output = self._create_expected_output(
            [
                "Three",
                " ",
                "times",
                "× ",
                "four",
                "^",
                "teen",
                " ÷",
                "divided",
                " ",
                "by",
                " [",
                "movies",
                "] ",
                "on",
                " \\",
                "slash",
                ".",
            ]
        )

        output = any_locale_word_tokenize(input_text)
        assert output == expected_output
```

**Review by XuesongYang — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**